### PR TITLE
Split Vision component to handle only one camera

### PR DESCRIPTION
**Pull request by auscompgeek**
This partially rewrites the Vision component to only handle localisation using a single camera.

Note to reviewers: using the hide whitespace option may make it easier to review this.

### CODE REVIEW

File: components/vision.py [hunk 1]
```diff
@@ -1,213 +1,161 @@
...
+        for target in results.getTargets():
+            poses = estimate_poses_from_apriltag(self.camera_to_robot, target)
+            if poses is None:
+                # tag doesn't exist
                 continue
-            self.last_timestamps[cam_idx] = timestamp
+            best_pose, alt_pose, self.last_pose_z = poses
+            pose = choose_pose(
+                best_pose,
+                alt_pose,
```
Comment (auscompgeek) [inline components/vision.py:79]: Every use of this poses tuple grabbed each of them by index. Name them for clarity.

File: components/vision.py [hunk 2]
```diff
@@ -1,213 +1,161 @@
...
-                    self.last_z_1 = z1
-                else:
-                    self.last_z_2 = z1
-                pose = choose_pose(
-                    poses[0],
-                    poses[1],
-                    self.chassis.get_pose(),
-                    target.getPoseAmbiguity(),
+            else:
+                self.rejected_in_row //= 2
```
Comment (auscompgeek) [inline components/vision.py:94]: Previously `rejected_in_row` would theoretically just go through exponential decay, never reaching 0, which was probably not intended.

File: components/vision.py [hunk 3]
```diff
@@ -1,213 +1,161 @@
...
+    add_to_estimator = tunable(False)
+    should_log = tunable(False)
+
+    rejected_in_row = tunable(0)
+    last_pose_z = tunable(0.0, writeDefault=False)
+
+    def __init__(
+        self,
+        # The name of the camera in PhotonVision.
+        name: str,
```
Comment (auscompgeek) [inline components/vision.py:30]: New feature in magicbot which I still haven't written docs for: it does dependency injection into the constructor :D

File: components/vision.py [hunk 4]
```diff
@@ -1,213 +1,161 @@
...
+        field: wpilib.Field2d,
+        data_log: wpiutil.log.DataLog,
+        chassis: Chassis,
+    ) -> None:
+        self.camera = PhotonCamera(name)
+        self.camera_to_robot = Transform3d(pos, rot).inverse()
+        self.last_timestamp = -1
+
+        self.field_pos_obj = field.getObject("vision_pose_" + name)
+        self.pose_log_entry = wpiutil.log.DoubleArrayLogEntry(
```
Comment (auscompgeek) [inline components/vision.py:44]: I presume logging only single-precision floating point numbers was a mistake.

**Review by james-ward — APPROVED** [review]
LGTM!

**Pull request merged**